text/0058-standard-library.md

```diff
@@ -0,0 +1,299 @@
+
+entity MyEntityType extends BaseEntityType { emailOptOut: Boolean }; // `MyEntityType` has all the attributes and parent types of `BaseEntityType`, and also has an attribute `emailOptOut` of type `Boolean`
+
+entity MyEntityType extends BaseEntityType in [MyGroupType] { emailOptOut: Boolean }; // combining the above examples
+```
+
+The base type may be an entity type, or may be a common type that aliases an
+entity type.
+As a consequence of this rule, chains of `extends` are allowed: you can write
+`entity B extends A` and later `entity C extends B`.
+
+To avoid cycles, the base type for `extends` must be a previously declared type
+(as in, declared farther up in the schema file).
+For this purpose, standard library declarations are considered to appear before
+line 1 of the schema file.
+If the base type is defined in a different schema file, that other schema file
+must be loaded first (e.g., must appear earlier in the iterator in [`Schema::from_schema_fragments()`]).
+
+## Drawbacks
+
+1. Cedar commits to maintaining all of the definitions in the Cedar standard
+library, which is a (hopefully small) maintenance burden. We can weigh this
+burden against the utility for each individual type declaration proposed for the
+Cedar standard library. This RFC only proposes the idea of the standard library,
+and a single definition `OIDC::User`.
+2. If a single declaration for `OIDC::User` (or other type) can't be agreed upon
+by the community, it may be difficult for Cedar to decide which definition
+should be in the standard library. Cedar may have to come up with best practices
+or tenets for resolving such disputes. We also could theoretically end up with a
+situation where part of the community uses the standard definition and another
+part uses a competing definition, which may or may not be better than how things
+would develop organically without this RFC.
+3. Users may incorrectly assume that `extends` has some kind of impact at
+runtime -- for instance that concrete values of `MyEntityType` inherit attribute
+values from a concrete value of `BaseEntityType` somehow.
+Hopefully we can address this with clear documentation.
+There's also nothing preventing us from providing API helpers that allow
+extending concrete `Entity` objects of a base type corresponding with a schema
+`extends` declaration (although this is not proposed in this RFC).
+4. Users may incorrectly assume that `extends` provides subtyping, while this
+RFC does not propose that it provides subtyping.
+For instance, users may assume that an action with principal type
+`BaseEntityType` could also be used with principal type `MyEntityType` which
+`extends` `BaseEntityType`; but this is not the case in the current proposal.
+It's also possible that the keyword being named `extends` in particular might
+make this problem worse due to existing associations with object-oriented
+inheritance; see Alternative H.
```

Comment:
This is my main gripe with the word `extends`. If we don't allow for this kind of subtyping then I think we should switch to `expands` (or similar).